core/state/trie_prefetcher.go

```diff
@@ -77,20 +97,62 @@ func newTriePrefetcher(db Database, root common.Hash, namespace string) *triePre
 		storageWasteMeter: metrics.GetOrRegisterMeter(prefix+"/storage/waste", nil),
 	}
 	go p.abortLoop()
+	go p.mainLoop()
 	return p
 }
 
-func (p *triePrefetcher) abortLoop() {
+func (p *triePrefetcher) mainLoop() {
 	for {
 		select {
-		case fetcher := <-p.abortChan:
-			fetcher.abort()
-		case <-p.closeChan:
-			// drain fetcher channel
+		case pMsg := <-p.prefetchChan:
+			fetcher := p.fetchers[pMsg.root]
+			if fetcher == nil {
+				fetcher = newSubfetcher(p.db, pMsg.root, pMsg.accountHash)
+				p.fetchersMutex.Lock()
+				p.fetchers[pMsg.root] = fetcher
+				p.fetchersMutex.Unlock()
+			}
+			fetcher.schedule(pMsg.keys)
+
+		case <-p.closeMainChan:
+			for _, fetcher := range p.fetchers {
+				p.abortChan <- fetcher // safe to do multiple times
+				<-fetcher.term
+				if metrics.EnabledExpensive {
+					if fetcher.root == p.root {
+						p.accountLoadMeter.Mark(int64(len(fetcher.seen)))
+						p.accountDupMeter.Mark(int64(fetcher.dups))
+						p.accountSkipMeter.Mark(int64(len(fetcher.tasks)))
+						fetcher.lock.Lock()
+						for _, key := range fetcher.used {
+							delete(fetcher.seen, string(key))
+						}
+						fetcher.lock.Unlock()
+						p.accountWasteMeter.Mark(int64(len(fetcher.seen)))
+					} else {
+						p.storageLoadMeter.Mark(int64(len(fetcher.seen)))
+						p.storageDupMeter.Mark(int64(fetcher.dups))
+						p.storageSkipMeter.Mark(int64(len(fetcher.tasks)))
+
+						fetcher.lock.Lock()
+						for _, key := range fetcher.used {
+							delete(fetcher.seen, string(key))
+						}
+						fetcher.lock.Unlock()
+						p.storageWasteMeter.Mark(int64(len(fetcher.seen)))
+					}
+				}
+			}
+			close(p.closeAbortChan)
+			close(p.closeMainDoneChan)
+			p.fetchersMutex.Lock()
+			p.fetchers = nil
+			p.fetchersMutex.Unlock()
+
+			// drain all the channels before quit the loop
 			for {
 				select {
-				case fetcher := <-p.abortChan:
-					fetcher.abort()
+				case <-p.prefetchChan:
```

Comment:
Do we need to explicitly close a channel? Since it will be GC once it is no referred any more.

Reply:
Yes, it will be collected by GC. But personally, I still habitually close or recycle unused resources manually.

This won't cause serious code problems, maybe can follow your style